src/cast_core/src/CommitLatency.cpp

```diff
@@ -0,0 +1,189 @@
+// Copyright 2019-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cast_core/actors/CommitLatency.hpp>
+
+#include <memory>
+
+#include <yaml-cpp/yaml.h>
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/json.hpp>
+#include <mongocxx/client.hpp>
+#include <mongocxx/collection.hpp>
+#include <mongocxx/database.hpp>
+
+#include <boost/log/trivial.hpp>
+#include <boost/throw_exception.hpp>
+
+#include <gennylib/Cast.hpp>
+#include <gennylib/ExecutionStrategy.hpp>
+#include <gennylib/InvalidConfigurationException.hpp>
+#include <gennylib/context.hpp>
+
+
+namespace genny::actor {
+
+struct CommitLatency::PhaseConfig {
+    mongocxx::collection collection;
+    mongocxx::write_concern wc;
+    mongocxx::read_concern rc;
+    mongocxx::read_preference rp;
+    std::string readModeString;
+    std::shared_ptr<mongocxx::client_session> session;
+    mongocxx::options::find optionsFind;
+    mongocxx::options::update optionsUpdate;
+    mongocxx::options::aggregate optionsAggregate;
+    mongocxx::options::transaction optionsTransaction;
+    bool useSession;
+    bool useTransaction;
+    int64_t repeat;
+    int64_t threads;
+    std::uniform_int_distribution<int64_t> amountDistribution;
+    ExecutionStrategy::RunOptions options;
+
+    PhaseConfig(PhaseContext& phaseContext, const mongocxx::database& db)
+        : collection{db[phaseContext.get<std::string>("Collection")]},
+          wc{phaseContext.get<mongocxx::write_concern>("WriteConcern")},
+          rc{phaseContext.get<mongocxx::read_concern>("ReadConcern")},
+          rp{phaseContext.get<mongocxx::read_preference>("ReadPreference")},
+          readModeString{phaseContext.get("ReadPreference")["ReadMode"].as<std::string>()},
+          useSession{phaseContext.get<bool, false>("Session").value_or(false)},
+          useTransaction{phaseContext.get<bool, false>("Transaction").value_or(false)},
+          repeat{phaseContext.get<IntegerSpec>("Repeat")},
+          threads{phaseContext.get<IntegerSpec>("Threads")},
+          amountDistribution{-100, 100},
+          options{ExecutionStrategy::getOptionsFrom(phaseContext, "ExecutionsStrategy")} {
+        if (useTransaction) {
+            optionsTransaction.write_concern(wc);
+            optionsTransaction.read_preference(rp);
+            // Transactions ignore the read_concern of a collection handle, must set transaction
```

Comment:
I'm not sure if it's helpful here, but you could also inherit the read_concern/write_concern/read_preference from the session's default_transaction_opts, or failing that, the client's.

See: https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#starttransaction

Reply:
Yes. I learned that the hard way when I first did this in pymongo. Since this test may or may not use transactions and sessions, I've tried to use only the options that are available in all permutations. I guess that wasn't fully possible anyway, but that was where I was coming from.